File: 백준/Gold/2064. IP 주소/IP 주소.js
const fs = require("fs");
const fileName = "input.txt";
const input = fs.readFileSync(0, "utf8").trim().split("\n");

const N = Number(input.shift());

function ipToBigInt(s) {
  const [a, b, c, d] = s.split(".").map(Number);
  return (
    (BigInt(a) << 24n) | (BigInt(b) << 16n) | (BigInt(c) << 8n) | BigInt(d)
  );
}

function bigIntToIp(x) {
  const a = Number((x >> 24n) & 255n);
  const b = Number((x >> 16n) & 255n);
  const c = Number((x >> 8n) & 255n);
  const d = Number(x & 255n);
  return `${a}.${b}.${c}.${d}`;
}

function prefixToMaskBigInt(prefix) {
  if (prefix === 0) return 0n;
  return ((1n << BigInt(prefix)) - 1n) << BigInt(32 - prefix);
}

let minIP = null,
  maxIP = null;
for (let i = 0; i < N; i++) {
  const ip = ipToBigInt(input[i]);
  if (minIP === null || ip < minIP) minIP = ip;
  if (maxIP === null || ip > maxIP) maxIP = ip;
}

let diff = minIP ^ maxIP; // diff = min XOR max
let prefix = 32;
if (diff !== 0n) {
  let msb = 0;
  while (diff > 0n) {
    diff >>= 1n;
    msb++;
  }
  // msb는 diff의 비트 길이(가장 오른쪽에 있는 1 위치+1)
  prefix = 32 - msb;
}

const mask =
  prefix === 0 ? 0n : ((1n << BigInt(prefix)) - 1n) << BigInt(32 - prefix);
const network = minIP & mask;

console.log(bigIntToIp(network));
console.log(bigIntToIp(mask));

File: 백준/Gold/2064. IP 주소/IP 주소.py
import sys

input = sys.stdin.readline


def bit_to_ip(b):
    res = []
    for i in range(4):
        res.append(str(b >> (8 * (3-i)) & 255))
    return ".".join(res)


n = int(input())

minIP = (1 << 32) - 1
maxIP = 0
for _ in range(n):
    ip = list(map(int, input().split(".")))
    bits = 0
    for i in range(3, -1, -1):
        bits |= ip[i] << (24-i*8)
    if bits < minIP:
        minIP = bits
    if bits > maxIP:
        maxIP = bits

ans = []
for m in range(33):
    mask = ((1 << 32) - 1) if m == 0 else ((1 << 32) - 1) ^ ((1 << m) - 1)
    if (minIP & mask) + (1 << m) > maxIP:
        ans.append(minIP & mask)
        ans.append(mask)
        break

for address in ans:
    print(bit_to_ip(address))